runtime-sdk/modules/contracts/src/abi/oasis/crypto.rs

```diff
@@ -121,6 +121,136 @@ impl<Cfg: Config> OasisV1<Cfg> {
+                                    }
+                                    crypto::x25519::Error::MalformedPrivateKey => {
+                                        Error::CryptoMalformedPrivateKey
+                                    }
+                                    crypto::x25519::Error::KeyDerivationFunctionFailure => {
+                                        Error::CryptoKeyDerivationFunctionFailure
+                                    }
+                                };
+                                ec.aborted = Some(err);
+                                wasm3::Trap::Abort
+                            })?;
+                        if output.len() != derived.len() {
+                            return Err(wasm3::Trap::Abort);
+                        }
+                        output.copy_from_slice(&derived);
+                        Ok(0)
+                    })?
+            },
+        );
+
+        #[allow(clippy::type_complexity)]
+        let deoxysii_factory =
+            |func: fn(&[u8], &[u8], &[u8], &[u8]) -> Result<Vec<u8>, crypto::deoxysii::Error>| {
+                move |ctx: wasm3::CallContext<'_, ExecutionContext<'_, C>>,
+                      (key, nonce, message, additional_data): (
+                    (u32, u32),
+                    (u32, u32),
+                    (u32, u32),
+                    (u32, u32),
+                )|
+                      -> Result<u32, wasm3::Trap> {
+                    // Make sure function was called in valid context.
+                    let ec = ctx.context.ok_or(wasm3::Trap::Abort)?;
+
+                    gas::use_gas(
+                        ctx.instance,
+                        ec.params.gas_costs.wasm_crypto_deoxysii_base
+                            + ec.params.gas_costs.wasm_crypto_deoxysii_byte
+                                * (message.1 as u64 + additional_data.1 as u64),
+                    )?;
+
+                    let output = ctx.instance.runtime().try_with_memory(
+                        |memory| -> Result<_, wasm3::Trap> {
+                            let key = Region::from_arg(key)
+                                .as_slice(&memory)
+                                .map_err(|_| wasm3::Trap::Abort)?;
+                            let nonce = Region::from_arg(nonce)
+                                .as_slice(&memory)
+                                .map_err(|_| wasm3::Trap::Abort)?;
+                            let message = Region::from_arg(message)
+                                .as_slice(&memory)
+                                .map_err(|_| wasm3::Trap::Abort)?;
+                            let additional_data = Region::from_arg(additional_data)
+                                .as_slice(&memory)
+                                .map_err(|_| wasm3::Trap::Abort)?;
+                            func(key, nonce, message, additional_data).map_err(|e| {
+                                let err = match e {
```

Comment:
This should only abort in case the input is obviously malformed (e.g. unexpected sizes), but not if there is a problem with authenticating/decrypting the data. In case of authentication errors, the error should be propagated so the caller can handle it.